### ISSUE TITLE
NuGet task fails on Push and Conflict in RMO: TypeError: Cannot read property 'replace' of undefined

### ISSUE THREAD

**Issue by abatishchev** [Microsoft/vsts-tasks#4824]
Hi,
I'm trying to replace old task NuGet Publisher with new task NuGet but it fails with Js/Ts error. I suspect a Conflict (attempt to push the same version of a package).

Log:
```
2017-07-19T00:11:53.6967405Z ##[section]Starting: Publish release package
2017-07-19T00:11:53.7123632Z ==============================================================================
2017-07-19T00:11:53.7123632Z Task         : NuGet
2017-07-19T00:11:53.7123632Z Description  : Restore, pack, or push NuGet packages, or run a NuGet command. Supports NuGet.org and authenticated feeds like Package Management and MyGet.
2017-07-19T00:11:53.7123632Z Version      : 2.0.5
2017-07-19T00:11:53.7123632Z Author       : Microsoft Corporation
2017-07-19T00:11:53.7123632Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613747)
2017-07-19T00:11:53.7123632Z ==============================================================================
2017-07-19T00:11:55.1030358Z Found tool in cache: NuGet 4.0.0 x64
2017-07-19T00:11:55.1030358Z Found tool in cache: NuGet 4.0.0 x64
2017-07-19T00:11:55.1186618Z Resolved from tool cache: 4.0.0
2017-07-19T00:11:55.1186618Z Using version: 4.0.0
2017-07-19T00:11:55.1186618Z Found tool in cache: NuGet 4.0.0 x64
2017-07-19T00:11:55.1186618Z [command]C:\Windows\system32\chcp.com 65001
2017-07-19T00:11:55.1186618Z Active code page: 65001
2017-07-19T00:11:55.1499366Z ##[error]TypeError: Cannot read property 'replace' of undefined
2017-07-19T00:11:55.1499366Z ##[error]Packages failed to publish
2017-07-19T00:11:55.1499366Z ##[section]Finishing: Publish release package
```

Internal url: https://microsoft.visualstudio.com/defaultcollection/Universal%20Store/Universal%20Store%20Team/_apps/hub/ms.vss-releaseManagement-web.hub-explorer?definitionId=2785&releaseId=196826&_a=release-logs
Internal alias: alexbat.

**Comment by petrformanek** [comment]
Hello,
we have the same issue today. It appeared all of a sudden during nuget restore.
Log:

> 2017-07-19T16:02:11.1902854Z ##[section]Starting: NuGet restore
2017-07-19T16:02:11.2059065Z ==============================================================================
2017-07-19T16:02:11.2059065Z Task         : NuGet
2017-07-19T16:02:11.2059065Z Description  : Restore, pack, or push NuGet packages, or run a NuGet command. Supports NuGet.org and authenticated feeds like Package Management and MyGet.
2017-07-19T16:02:11.2059065Z Version      : 2.0.5
2017-07-19T16:02:11.2059065Z Author       : Microsoft Corporation
2017-07-19T16:02:11.2059065Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613747)
2017-07-19T16:02:11.2059065Z ==============================================================================
2017-07-19T16:02:11.8309134Z Found tool in cache: NuGet 4.0.0 x64
2017-07-19T16:02:11.8465352Z Found tool in cache: NuGet 4.0.0 x64
2017-07-19T16:02:11.8465352Z Resolved from tool cache: 4.0.0
2017-07-19T16:02:11.8465352Z Using version: 4.0.0
2017-07-19T16:02:11.8465352Z Found tool in cache: NuGet 4.0.0 x64
2017-07-19T16:02:11.8621588Z [command]C:\Windows\system32\chcp.com 65001
2017-07-19T16:02:11.8621588Z Active code page: 65001
2017-07-19T16:02:11.9715360Z Detected NuGet version 4.0.0.2283 / 4.0.0
2017-07-19T16:02:11.9715360Z SYSTEMVSSCONNECTION exists true
2017-07-19T16:02:11.9871695Z ##[error]TypeError: Cannot read property 'replace' of undefined
2017-07-19T16:02:11.9871695Z ##[error]Packages failed to restore
2017-07-19T16:02:11.9871695Z ##[section]Finishing: NuGet restore

**Comment by abatishchev** [comment]
Any response on this, please? We're blocked by this bug actually.

**Comment by petrformanek** [comment]
What worked for me was to clone the build configuration. The cloned one doesn't have such issue.
Strange but works for me.
Hope it will help you too.

**Comment by abatishchev** [comment]
I haven't figured out the conditions when this exception occurs. Are you saying that your cloned definition works in 100% cases?

**Comment by petrformanek** [comment]
I was executing only a couple of builds and all were without issue.

**Comment by abatishchev** [comment]
Any response on this, please?

**Comment by abatishchev** [comment]
Any response on this, please?

**Comment by bryanmacfarlane** [comment]
Adding @keithrob from packaging team

**Comment by keithrob** [comment]
@abatishchev & @petrformanek ,
Two things are happening here.  

1. The first is that you're using "External NuGet server" incorrectly.  You should click the manage button and create a NuGet service endpoint with the requisite credentials.  Then the you'll be able to pick that endpoint from the dropdown.  It looks like you're using an ENVAR, i.e. $(NuGetFeed),  which isn't what that selection box was intended for.
2. We have a bug that allows you to type arbitrary text into that box.  You should **only** be able to pick service endpoints. 

We'll fix 2.  However, in the meantime if you fix 1 you should be unblocked.

**Comment by abatishchev** [comment]
@keithrob thanks for looking into this!

When I try to add an endpoint (I use the same VSTS), it asks for an url and a personal token. So here's few thoughts:
- Can you add another option saying "Current VSTS instance", I believe it would be quite a popular choice (alongside with nuget.org)
- Can it authenticate by itself, or in other way do not require user token at build server?

**Comment by keithrob** [comment]
If you want to push to a feed in the current account use "This account/collection"

**Comment by abatishchev** [comment]
Right, I was going to update/delete my comment with same :) Thanks again!

**Comment by abatishchev** [comment]
However please add a search-as-you-type capability to the dropdown where you select a feed

**Comment by abatishchev** [comment]
Here's a defect I found:
1. Create an endpoint
2. Select the endpoint, save the (release) definition
3. Delete the endpoint
4. Try to change from an endpoint to "this account/collection", observe an error:

>VS402913: The release definition cannot be saved because the environment 'Pre-release' references a service endpoint that cannot be found, or that you do not have permission to access. Details: 'Task 'Publish pre-release package' is using endpoint 'ef6d05df-236e-44e0-94cc-c60f409f8b44''

**Comment by keithrob** [comment]
@abatishchev ,
Please try switching to the new "Release Definition Editor" under "Preview Features".  This should be fixed there.

**Comment by abatishchev** [comment]
This happened to me in the new editor

**Comment by keithrob** [comment]
@GitHubSriramB Please see last 4 comments.

**Comment by GitHubSriramB** [comment]
@abatishchev - Reg. the sequence that leads to the error when saving release definition - are you sure there are no usage of "Task Publish pre-release package" in "Pre-release" environment where the endpoint that you deleted is still being used?

**Comment by abatishchev** [comment]
Hi Shiram,
I deleted an endpoint in one tab and had release definition editor opened in another tab. Once deleted, I switched over the tabs, selected another radio button (this account/collection) while having not existing anymore endpoint under another radio button, repeated this for both environments. To me it seems that it was validating the endpoints even that option was not selected. Once I selected it back, deleted all text, selected back again then was able to save just fine.

**Comment by GitHubSriramB** [comment]
Got it. Yes, I believe the endpoint is still referred by task although you choose the other radio button. We are aware of this issue & will fix it in a future release. Until then the workaround is to clear the selection & save as you have done.
